### PR TITLE
Android Update Without Appcast

### PR DESCRIPTION
**Pull request by Williano**
1. Android Without Appcast
I have implemented a functionality that will allow Android updates without Appcast. The feature will pull the App's latest version and release notes from the PlayStore just like how the iTunes version of the iOS is implemented. It also allow developers to optionally  enter the Android ID of their apps in UpgradeAlert and UpgradeCard if it is different from what is currently in their app. The only package I added was dart html to parse the html I am downloading from the PlayStore.

2. AlertDialog Buttons
I have changed the UpdateNow button to an ElevatedButton since is a call to action button and must be different from Ignore and Later.

Screenshot_20210628-041918

### CODE REVIEW

File: lib/src/upgrader.dart [hunk 1]
```diff
@@ -519,7 +567,7 @@ class Upgrader {
           TextButton(
               child: Text(messages!.message(UpgraderMessage.buttonTitleLater)!),
               onPressed: () => onUserLater(context, true)),
-        TextButton(
+        ElevatedButton(
```
Comment (larryaasen) [inline lib/src/upgrader.dart:570]: Can you show me in the Material Design spec where this applies?
https://material.io/components/dialogs#anatomy

File: lib/src/upgrader.dart [hunk 2]
```diff
@@ -497,7 +542,10 @@ class Upgrader {
           ));
     }
     return AlertDialog(
-      title: Text(title),
+      title: Text(
+        title,
+        textAlign: TextAlign.center,
```
Comment (larryaasen) [inline lib/src/upgrader.dart:547]: Can you show me in the Material Design spec where this applies?
https://material.io/components/dialogs#anatomy

File: lib/src/upgrader.dart [hunk 3]
```diff
@@ -239,6 +250,40 @@ class Upgrader {
...
+    final uri =
+        Uri.https('play.google.com', '/store/apps/details', {'id': '$id'});
+    final response = await http.get(uri);
+    if (response.statusCode != 200) {
+      debugPrint('Can\'t find an app in the Play Store with the id: $id');
+      return null;
+    }
+    final document = parse(response.body);
+
+    final additionalInfoElements = document.getElementsByClassName('hAyfc');
```
Comment (larryaasen) [inline lib/src/upgrader.dart:265]: Is there an API contract that guarantees that `hAyfc` will always be available?

File: lib/src/upgrader.dart [hunk 4]
```diff
@@ -44,6 +45,11 @@ class Upgrader {
   /// When an appcast is configured for iOS, the iTunes lookup is not used.
   AppcastConfiguration? appcastConfig;
 
+  /// An optional value that can override the default packageName when
+  /// attempting to reach the Google Play Store. This is useful if your app has
+  /// a different package name in the Play Store.
```
Comment (larryaasen) [inline lib/src/upgrader.dart:50]: Can you provide an example when the package name does not match the Play Store?

File: lib/src/upgrade_messages.dart [hunk 5]
```diff
@@ -161,7 +161,7 @@ class UpgraderMessages {
       case 'en':
       default:
         message =
-            'A new version of {{appName}} is available! Version {{currentAppStoreVersion}} is now available-you have {{currentInstalledVersion}}.';
+            'A new version of {{appName}} is available! Version {{currentAppStoreVersion}} is now available - you have {{currentInstalledVersion}}.';
```
Comment (larryaasen) [inline lib/src/upgrade_messages.dart:164]: This change is not related to Android and the Play Store.

File: lib/src/upgrader.dart (hunk 1 above)
Comment (Williano) [reply lib/src/upgrader.dart:570]: No please

File: lib/src/upgrader.dart (hunk 2 above)
Comment (Williano) [reply lib/src/upgrader.dart:547]: This for my personal project and should not have added it to the PR

File: lib/src/upgrade_messages.dart (hunk 5 above)
Comment (Williano) [reply lib/src/upgrade_messages.dart:164]: Yes, it was for my project and should not have added it

File: lib/src/upgrader.dart (hunk 4 above)
Comment (Williano) [reply lib/src/upgrader.dart:50]: The android ID is asked as an optional value and will be used Incase the package name does not exist.

final id = androidId ??_packageInfo!.packageName;

File: lib/src/upgrader.dart (hunk 3 above)
Comment (Williano) [reply lib/src/upgrader.dart:265]: From my research, Google doesn't change their DOM ID's often and this has been the ID for years  now.

https://stackoverflow.com/questions/34309564/how-to-get-app-market-version-information-from-google-play-store

File: lib/src/upgrade_messages.dart (hunk 5 above)
Comment (Williano) [reply lib/src/upgrade_messages.dart:164]: I will change it back to the what you originally did

File: lib/src/upgrader.dart (hunk 2 above)
Comment (Williano) [reply lib/src/upgrader.dart:547]: I will change it back to what you originally did

File: lib/src/upgrader.dart (hunk 1 above)
Comment (Williano) [reply lib/src/upgrader.dart:570]: I will change it back to what you originally did

Comment (Williano) [reply lib/src/upgrader.dart]: I just made a new commit to reset it to the default

File: lib/src/upgrader.dart (hunk 2 above)
Comment (Williano) [reply lib/src/upgrader.dart]: I just made a new commit to reset it to the default

File: lib/src/upgrader.dart (hunk 1 above)
Comment (Williano) [reply lib/src/upgrader.dart]: Screenshot_20210630-011154

File: lib/src/upgrade_messages.dart (hunk 5 above)
Comment (Williano) [reply lib/src/upgrade_messages.dart:164]: I just made a new commit to reset it to the default

**Comment by Williano** [comment]
I will be on standby in case you want me to make any changes.

Thank you.

File: lib/src/upgrader.dart (hunk 4 above)
Comment (larryaasen) [reply lib/src/upgrader.dart:50]: When packageName does exist, should packageName be used instead of `androidId`? Can you show me an example of when packageName does not exist for an Android app?

File: lib/src/upgrader.dart [hunk 6]
```diff
@@ -239,6 +250,40 @@ class Upgrader {
     return true;
   }
 
+  /// Android info is fetched by parsing the html of the app store page.
+  Future<bool?> _getAndroidStoreVersion() async {
```
Comment (larryaasen) [inline lib/src/upgrader.dart:254]: Fetching the Android version from the Google Play Store should be in its own class and file. Use `ITunesSearchAPI` as an example. The call to fetch the version from the Google Play Store should happen after the appcast, just like `ITunesSearchAPI`. It should happen right after the call to `findCountryCode()`. The logic for this new code needs to align better with the existing logic in `_updateVersionInfo`.

File: lib/src/upgrade_messages.dart (hunk 5 above)
Comment (larryaasen) [inline lib/src/upgrade_messages.dart:164]: This change is not related to Android and the Play Store.

File: lib/src/upgrader.dart (hunk 6 above)
Comment (Williano) [reply lib/src/upgrader.dart:254]: Okay, I will work on that and let you know when I finish.

File: lib/src/upgrade_messages.dart (hunk 5 above)
Comment (Williano) [reply lib/src/upgrade_messages.dart:164]: I will fix the change for this also

**Comment by Williano** [comment]
I am sorry you have not heard back from me yet, I’ll make a pull request with the changes you asked for this week.

**Comment by Williano** [comment]
Hi Mr. Larry,

I have refactored the code base and added tests to test the two classes and their implementation.

I am on standby. 

Thank you.

**Comment by bdlukaa** [comment]
@larryaasen Hello! Any chances of this being reviewed/merged soon? It's such a great package, but needing Appcast is the only thing blocking me from using it

File: lib/src/upgrade_messages.dart (hunk 5 above)
Comment (larryaasen) [reply lib/src/upgrade_messages.dart:164]: @Williano Check again.

File: lib/src/play_store_search_api.dart [hunk 7]
```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2021 William Kwabla. All rights reserved.
+ */
+
+import 'package:html/parser.dart' show parse;
+import 'package:http/http.dart' as http;
+
+class PlayStroeSearchApi {
```
Comment (larryaasen) [inline lib/src/play_store_search_api.dart:8]: @Williano The word stroe should be store.

Comment (larryaasen) [reply lib/src/play_store_search_api.dart:8]: To follow the same conventions as `ITunesSearchAPI`, it should be named `PlayStoreSearchAPI`.

File: lib/src/play_store_search_api.dart [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+  /// Provide an HTTP Client that can be replaced for mock testing.
+  http.Client? client = http.Client();
+
+  bool debugEnabled = false;
+
+  /// Look up by id.
+  /// Example: look up Kotoko Express Android App:
+  /// ```lookupURLById('com.kotoko.express');```
+
```
Comment (larryaasen) [inline lib/src/play_store_search_api.dart:20]: Remove extra line.

File: lib/src/play_store_search_api.dart [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+import 'package:html/parser.dart' show parse;
+import 'package:http/http.dart' as http;
+
+class PlayStroeSearchApi {
+  /// Play Store Search Api URL
+  final String playStorePrefixURL = 'play.google.com';
+
+  /// Provide an HTTP Client that can be replaced for mock testing.
+  http.Client? client = http.Client();
```
Comment (larryaasen) [inline lib/src/play_store_search_api.dart:13]: `client` was never used.

File: lib/src/play_store_search_api.dart [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  Uri? lookupURLById(id) {
+    final uri =
+        Uri.https(playStorePrefixURL, '/store/apps/details', {'id': '$id'});
+
+    return uri;
+  }
+
+  dynamic decodeResults(jsonResponse) {
```
Comment (larryaasen) [inline lib/src/play_store_search_api.dart:45]: Private methods should start with `_`.

Comment (larryaasen) [reply lib/src/play_store_search_api.dart:45]: This method should return a type instead of `dynamic`. Is this a `Map`?

File: lib/src/upgrader.dart [hunk 11]
```diff
@@ -10,8 +10,10 @@ import 'package:flutter/material.dart';
 import 'package:http/http.dart' as http;
 import 'package:package_info/package_info.dart';
 import 'package:shared_preferences/shared_preferences.dart';
+import 'package:upgrader/src/play_store_search_api.dart';
 import 'package:url_launcher/url_launcher.dart';
 import 'package:version/version.dart';
+// import 'package:html/parser.dart' show parse;
```
Comment (larryaasen) [inline lib/src/upgrader.dart:16]: Remove comment.

File: lib/src/upgrader.dart [hunk 12]
```diff
@@ -10,8 +10,10 @@ import 'package:flutter/material.dart';
...
 
 import 'appcast.dart';
 import 'itunes_search_api.dart';
```
Comment (larryaasen) [inline lib/src/upgrader.dart:19]: Should include play_store_search_api.dart after this line.

File: lib/src/upgrader.dart [hunk 13]
```diff
@@ -44,6 +46,11 @@ class Upgrader {
   /// When an appcast is configured for iOS, the iTunes lookup is not used.
   AppcastConfiguration? appcastConfig;
 
+  /// An optional value that can override the default packageName when
+  /// attempting to reach the Google Play Store. This is useful if your app has
+  /// a different package name in the Play Store.
+  String? androidId;
```
Comment (larryaasen) [inline lib/src/upgrader.dart:52]: A better name for this variable would be `applicationId`, since that is what the Android docs use. See https://developer.android.com/studio/build/application-id

File: lib/src/upgrader.dart [hunk 14]
```diff
@@ -223,6 +230,11 @@ class Upgrader {
         print('upgrader: countryCode: $code');
       }
 
+      // Get android update without appcast
+      if (Platform.isAndroid) {
```
Comment (larryaasen) [inline lib/src/upgrader.dart:234]: This line would not execute because of line 219:
```
      if (Platform.isAndroid) {
        return false;
      }
```

**Review by larryaasen — CHANGES_REQUESTED** [review]
Still have a few things to cleanup.

**Comment by Williano** [comment]
Okay I will fix the changes and get back to you

**Review by Williano — COMMENTED** [review]
I have fixed all the issues you raised.

**Comment by Williano** [comment]
I will be on standy for any changes.

Thank you.

**Comment by ajviera** [comment]
I hope you can merge this quickly, it is just what I need to be able to deliver this functionality.

**Comment by developerflutter** [comment]
Can we have this functionality live ?

**Comment by bdlukaa** [comment]
@larryaasen

File: lib/src/play_store_search_api.dart [hunk 15]
```diff
@@ -2,22 +2,16 @@
  * Copyright (c) 2021 William Kwabla. All rights reserved.
  */
 
+import 'package:html/dom.dart';
 import 'package:html/parser.dart' show parse;
 import 'package:http/http.dart' as http;
 
-class PlayStroeSearchApi {
+class PlayStroeSearchAPI {
```
Comment (larryaasen) [inline lib/src/play_store_search_api.dart:9]: The spelling of this class should be `PlayStoreSearchAPI`.

File: lib/src/play_store_search_api.dart [hunk 16]
```diff
@@ -2,22 +2,16 @@
...
   /// Play Store Search Api URL
   final String playStorePrefixURL = 'play.google.com';
 
-  /// Provide an HTTP Client that can be replaced for mock testing.
-  http.Client? client = http.Client();
```
Comment (larryaasen) [inline lib/src/play_store_search_api.dart]: By removing `client` you are no longer able to unit test with a mock. I think this should be reverted. See class `ITunesSearchAPI` for an example.

File: lib/src/play_store_search_api.dart [hunk 17]
```diff
@@ -2,22 +2,16 @@
...
-
   bool debugEnabled = false;
 
-  /// Look up by id.
```
Comment (larryaasen) [inline lib/src/play_store_search_api.dart]: Why did you remove these comments? Normally, comments can be helpful.

File: test/play_store_test.dart [hunk 18]
```diff
@@ -5,38 +5,35 @@
...
     expect(
-        playStore.lookupURLById(androidId),
+        playStore.lookupURLById(applicationId),
         equals(Uri.https(
-            'play.google.com', '/store/apps/details', {'id': androidId})));
+            'play.google.com', '/store/apps/details', {'id': applicationId})));
   });
 
   test('testing lookupById', () async {
-    final client = MockPlayStoreSearchClient.setupMockClient();
```
Comment (larryaasen) [inline test/play_store_test.dart]: Why did you remove the use of `MockPlayStoreSearchClient ` in this test?

File: lib/src/upgrader.dart (hunk 4 above)
Comment (larryaasen) [reply lib/src/upgrader.dart:50]: I don't think this is possible. Can you show me an example of when packageName does not exist for an Android app?

File: lib/src/play_store_search_api.dart (hunk 15 above)
Comment (Williano) [reply lib/src/play_store_search_api.dart:9]: I will fix it ASAP

File: lib/src/play_store_search_api.dart (hunk 16 above)
Comment (Williano) [reply lib/src/play_store_search_api.dart]: I will add it back

File: lib/src/play_store_search_api.dart (hunk 17 above)
Comment (Williano) [reply lib/src/play_store_search_api.dart]: I will add it back

File: test/play_store_test.dart (hunk 18 above)
Comment (Williano) [reply test/play_store_test.dart]: Ok

**Comment by Williano** [comment]
I have fixed all the changes now. 

I will be on stand-by.

File: lib/src/upgrade_messages.dart (hunk 5 above)
Comment (larryaasen) [inline lib/src/upgrade_messages.dart:164]: Please revert this change: `available - you`

File: lib/src/upgrader.dart (hunk 4 above)
Comment (larryaasen) [reply lib/src/upgrader.dart:50]: @Williano I don't think this is possible. Can you show me an example of when packageName does not exist for an Android app?

File: lib/src/upgrader.dart [hunk 19]
```diff
@@ -223,6 +224,12 @@ class Upgrader {
         print('upgrader: countryCode: $code');
       }
 
+      // If this platform is not iOS, skip the iTunes lookup
+      // Get android update without appcast
+      if (Platform.isAndroid) {
+        await _getAndroidStoreVersion();
+      }
+
       final iTunes = ITunesSearchAPI();
```
Comment (larryaasen) [inline lib/src/upgrader.dart:233]: Why does ITunesSearchAPI still get called for Android? There is no need to use ITunesSearchAPI on Android. This logic should have an else if Platform.isIOS.

File: lib/src/upgrader.dart [hunk 20]
```diff
@@ -10,6 +10,7 @@ import 'package:flutter/material.dart';
 import 'package:http/http.dart' as http;
 import 'package:package_info/package_info.dart';
 import 'package:shared_preferences/shared_preferences.dart';
+import 'package:upgrader/src/play_store_search_api.dart';
```
Comment (larryaasen) [inline lib/src/upgrader.dart:13]: @Williano Local files are declared down a few lines.

File: lib/src/upgrader.dart [hunk 21]
```diff
@@ -239,6 +246,23 @@ class Upgrader {
     return true;
   }
 
+  /// Android info is fetched by parsing the html of the app store page.
+  Future<bool?> _getAndroidStoreVersion() async {
+    final id = applicationId ?? _packageInfo!.packageName;
+
+    final PlayStore = PlayStoreSearchAPI();
```
Comment (larryaasen) [inline lib/src/upgrader.dart:253]: @Williano Local variables in Dart do not start with upper case.

File: test/mock_play_store_client.dart [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+    var id = 'com.kotoko.express';
+
+    final client = MockClient((http.Request request) async {
+      final response =
+          '{"results": [{"version": "2.1.6", "id": "com.kotoko.express",  "releaseNotes": "Bug fixes and performance enhancements"}]}';
+
+      final url = request.url.toString();
+
+      // ignore: unrelated_type_equality_checks
+      if (url == PlayStoreSearchAPI().lookupById(id)) {
```
Comment (larryaasen) [inline test/mock_play_store_client.dart:22]: @Williano `lookupById` returns `Future<Document?>` but this is comparing to a String url. That does not make sense. Why did you ignore unrelated_type_equality_checks?

File: lib/src/play_store_search_api.dart [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+
+  /// Look up by id.
+  Future<Document?> lookupById(String id) async {
+    if (id.isEmpty) {
+      return null;
+    }
+
+    final url = lookupURLById(id)!;
+
+    final response = await http.get(url);
```
Comment (larryaasen) [inline lib/src/play_store_search_api.dart:26]: @Williano Instead of using `http` here, it should use `client`, so it can be mocked.

**Comment by larryaasen** [comment]
@Williano The `flutter test` command in the `.travis.yml` file was not working as expected, and your unit tests were not being run on Travis CI. You can see that here: https://app.travis-ci.com/github/larryaasen/upgrader/builds/234547428

Please add this line to the `.travis.yml` file so that your unit tests in `play_store_test.dart` will run.
`flutter test test/play_store_test.dart`

Add this after the line: `flutter test --coverage --coverage-path=lcov.info`

File: lib/src/play_store_search_api.dart (hunk 23 above)
Comment (Williano) [reply lib/src/play_store_search_api.dart]: When I use the client!.get(url), it returns null but when I use the http.get(url) it returns the HTML content.

**Comment by Williano** [comment]
> @Williano The `flutter test` command in the `.travis.yml` file was not working as expected, and your unit tests were not being run on Travis CI. You can see that here: https://app.travis-ci.com/github/larryaasen/upgrader/builds/234547428
> 
> Please add this line to the `.travis.yml` file so that your unit tests in `play_store_test.dart` will run.
> `flutter test test/play_store_test.dart`
> 
> Add this after the line: `flutter test --coverage --coverage-path=lcov.info`

Please should this be before or after the coverage line?

**Comment by Williano** [comment]
Adding it after the coverage makes all the test fail

**Review by larryaasen — CHANGES_REQUESTED** [review]

File: test/play_store_test.dart [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+  test('testing lookupById', () async {
+    final client = MockPlayStoreSearchClient.setupMockClient();
+    final playStore = PlayStoreSearchAPI();
+    playStore.client = client;
+
+    final response = await playStore.lookupById(applicationId);
+    expect(response, isInstanceOf<Document>());
+
+    final results = response!;
+    expect(results, isNotNull);
```
[the remark this answers is not in the document]
Comment (Williano) [reply test/play_store_test.dart:35]: Thank you for the feedback. I will remove it and open a New PR

**Comment by Williano** [comment]
Thank you for merging my pull request and allowing me to contribute to this package.

If you want me to update the document with this new PR, I can do that also.


Thank you

**Comment by larryaasen** [comment]
I will take care of that. More testing is required.

**Comment by Williano** [comment]
Okay

Thank you.

**Comment by hicnar** [comment]
@Williano thanks for implementing this! @larryaasen any time estimates as to when the package including this functionality  be released to pub.dev?

File: lib/src/play_store_search_api.dart [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+    final releaseNotes = releaseNotesElement
+        .querySelector('.PHBdkd')
+        ?.querySelector('.DWPxHb')
+        ?.text;
+
+    return releaseNotes;
+  }
+
+  /// Return field trackViewUrl from Play Store results.
+  static String? trackViewUrl(id) {
```
Comment (larryaasen) [inline lib/src/play_store_search_api.dart:72]: @Williano This method `trackViewUrl` does not return anything from the Play Store results. What is this for? It is no different from `lookupURLById`.

File: lib/src/play_store_search_api.dart [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  /// Return field version from Play Store results.
+  static String? version(response) {
+    final additionalInfoElements = response.getElementsByClassName('hAyfc');
+    final versionElement = additionalInfoElements.firstWhere(
+      (elm) => elm.querySelector('.BgcNfc')!.text == 'Current Version',
+    );
+
+    final storeVersion = versionElement.querySelector('.htlgb')!.text;
```
Comment (larryaasen) [inline lib/src/play_store_search_api.dart:87]: @Williano This does not work for app `com.google.android.apps.maps`. It returns `Varies with device`.

**Comment by larryaasen** [comment]
@Williano I added the `bin/playstore_loopup.dart` app for testing. Usage from the terminal:
```
dart playstore_lookup.dart id=com.google.android.apps.maps
```

**Comment by larryaasen** [comment]
> @larryaasen any time estimates as to when the package including this functionality be released to pub.dev?

@hicnar There are still issues to resolve.

File: lib/src/play_store_search_api.dart (hunk 25 above)
Comment (Williano) [reply lib/src/play_store_search_api.dart:72]: Yes, is not different from lookupURLById

File: lib/src/play_store_search_api.dart (hunk 26 above)
Comment (Williano) [reply lib/src/play_store_search_api.dart:87]: Most of the google apps and some apps doesn’t show current version on play store because Google Play allows you to publish different APKs for your application. Each one targeted to different device configuration. Thus, each APK is an independent version of your application, but they share the same application listing on Google Play and must share the same package name and be signed with the same release key.

Usually Android applications run on most compatible devices with a single APK, by supplying alternative resources for different configurations (e.g., different layouts for different screen sizes) and the Android system selects the appropriate resources for the device at runtime. But in a few cases, however, a single APK is unable to support all device configurations, because alternative resources make the APK file too big (greater than 50MB) or other technical challenges prevent a single APK from working on all devices.

**And even though Google doesn't encourage developers to use this "multiple APK" feature, most Google apps use it: Google+, Google Chrome, Gmail, Google Maps, Google Search, Google Play Books, Google Play Movies & TV, Google Translate**

**Comment by larryaasen** [comment]
I have resolved all of these related issues and published version 3.5.0 with this PR and other updates. Thanks to @Williano for his work on this.

**Comment by Williano** [comment]
Wow, thank you very much @larryaasen for letting me contribute to this project. 

I really appreciate it.
